cdap-security/src/main/java/io/cdap/cdap/security/auth/TinkCipher.java

```diff
@@ -0,0 +1,118 @@
+import com.google.crypto.tink.JsonKeysetReader;
+import com.google.crypto.tink.KeysetHandle;
+import com.google.crypto.tink.aead.AeadConfig;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.conf.SConfiguration;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.Base64;
+
+/**
+ * Tink cipher that allows encrypting and decrypting data using keyset stored in {@link SConfiguration}
+ */
+public class TinkCipher {
+  /**
+   * Wraps a keyset with some additional parameters and metadata.
+   */
+  private final KeysetHandle keysetHandle;
+
+  /**
+   * Tink AEAD (Authenticated Encryption with Associated Data) primitive
+   */
+  private final Aead aead;
+
+  public TinkCipher(SConfiguration sConf) throws CipherException {
+    try {
+      // Init Tink with AEAD primitive
+      AeadConfig.register();
+
+      // Load keyset from sConf
+      String jsonKeyset = sConf.get(Constants.Security.Authentication.USER_CREDENTIAL_ENCRYPTION_KEYSET);
+      this.keysetHandle = CleartextKeysetHandle.read(JsonKeysetReader.withString(jsonKeyset));
+
+      this.aead = keysetHandle.getPrimitive(Aead.class);
+    } catch (GeneralSecurityException e) {
+      throw new CipherException("Failed to init Tink cipher", e);
+    } catch (IOException e) {
+      throw new CipherException("Failed to load Tink keyset", e);
+    }
+  }
+
+  /**
+   * Encrypt the data.
+   *
+   * @param plainData data to be encrypted
+   * @param associatedData used for integrity checking during decryption.
+   * @return encrypted data
+   * @throws CipherException if encryption fails
+   */
+  public byte[] encrypt(byte[] plainData, byte[] associatedData) throws CipherException {
+    try {
+      return aead.encrypt(plainData, associatedData);
+    } catch (GeneralSecurityException e) {
+      throw new CipherException("Failed to encrypt", e);
+    }
+  }
+
+  /**
+   * Encrypt the data and return encrypted data in base64 encoded form.
+   *
+   * @param plainData data to be encrypted
+   * @param associatedData used for integrity checking during decryption.
+   * @return encrypted data in base64 encoded form
+   * @throws CipherException if encryption fails
+   */
+  public String encryptToBase64(byte[] plainData, byte[] associatedData) throws CipherException {
+    return Base64.getEncoder().encodeToString(encrypt(plainData, associatedData));
+  }
+
+  /**
+   * Decrypt the cipher data.
+   *
+   * @param cipherData data to be decrypted
+   * @param associatedData used for integrity checking, must be the same as that used during encryption.
+   * @return decrypted data
+   * @throws CipherException if decryption fails
+   */
+  public byte[] decrypt(byte[] cipherData, byte[] associatedData) throws CipherException {
+    try {
+      return aead.decrypt(cipherData, associatedData);
+    } catch (GeneralSecurityException e) {
+      throw new CipherException("Failed to decrypt", e);
```

Comment:
May be good to add exception message and some more text on what kind of encryption/decryption failed

Reply:
Done